File: caifuclaw_business_app/frontend/src/api/trafficAnalytics.ts
/**
 * Company: 深圳智柠网络科技有限公司
 * Author: mohsen liang
 */

import { get, post } from './http'

export type TrafficMetricKey = 'impressions' | 'clicks' | 'add_to_cart' | 'orders'
export type TrafficRateMetricKey = 'ctr' | 'cvr'
export type TrafficRankingMetricKey = TrafficMetricKey | TrafficRateMetricKey
export type TrafficSortOrder = 'asc' | 'desc'
export type TrafficCoverage = 'full' | 'partial' | 'unavailable'
export type TrafficComparisonDimension = 'sku' | 'category'
export type TrafficComparisonSort =
  | 'delta_abs'
  | 'rate_desc'
  | 'rate_asc'
  | 'current_desc'
  | 'current_asc'
  | 'previous_desc'
  | 'previous_asc'
  | 'delta_desc'
  | 'delta_asc'
export type TrafficComparisonChangeDirection = 'all' | 'up' | 'down' | 'flat'

export interface TrafficCapability {
  label: string
  scope: string
  grain: string
  metrics: string[]
  note: string
}

export interface TrafficSyncRun {
  id: number
  platform_account_id: number
  platform: string
  account_id: string
  shop_name: string
  status: 'pending' | 'running' | 'success' | 'partial_success' | 'failed' | 'timed_out' | string
  date_from: string
  date_to: string
  rows_written: number
  error_message: string
  triggered_by: string
  started_at?: string | null
  finished_at?: string | null
  created_at?: string | null
}

export interface TrafficAccount {
  id: number
  platform: string
  account_id: string
  display_name: string
  enabled: boolean
  authorization_status: string
  capability: TrafficCapability
  latest_run?: TrafficSyncRun | null
  latest_metric_at?: string | null
  latest_period_start?: string | null
  latest_period_end?: string | null
  data_freshness: 'fresh' | 'stale' | 'missing'
}

export interface TrafficMetricRow {
  rank?: number
  platform: string
  platform_account_id: number
  account_id: string
  shop_name: string
  source: string
  grain: string
  stat_date?: string
  region: string
  entity_type?: string
  entity_id?: string
  sku?: string
  product_name?: string
  impressions: number | null
  clicks: number | null
  add_to_cart: number | null
  orders: number | null
  buyers: number | null
  units_sold: number | null
  negative_reviews: number | null
  revenue: number | null
  sales_share?: number | null
  currency: string
  ctr: number | null
  cart_rate: number | null
  cvr: number | null
  cart_conversion: number | null
  period_start: string
  period_end: string
  synced_at: string
  coverage: Record<string, TrafficCoverage>
}

export interface TrafficCategoryRow extends TrafficMetricRow {
  platform_category_id: string
  platform_category_name: string
  platform_category_path: string
  categorized: boolean
  sku_count: number
}

export interface TrafficComparisonRow {
  rank: number
  platform: string
  platform_account_id: number
  account_id: string
  shop_name: string
  source: string
  region: string
  grain?: string
  entity_type?: string
  entity_id?: string
  sku?: string
  product_name?: string
  platform_category_id?: string
  platform_category_name?: string
  platform_category_path?: string
  current_impressions: number | null
  previous_impressions: number | null
  delta_impressions: number | null
  delta_rate_impressions: number | null
  current_clicks: number | null
  previous_clicks: number | null
  delta_clicks: number | null
  delta_rate_clicks: number | null
  current_add_to_cart: number | null
  previous_add_to_cart: number | null
  delta_add_to_cart: number | null
  delta_rate_add_to_cart: number | null
  current_orders: number | null
  previous_orders: number | null
  delta_orders: number | null
  delta_rate_orders: number | null
}

export type TrafficSkuFocusReason =
  | 'high_impressions_no_orders'
  | 'high_clicks_missing_impressions_or_cart'
  | 'high_cart_missing_orders'
  | 'high_orders_missing_impressions'

export interface TrafficSkuFocusRow {
  rank: number
  platform: string
  platform_account_id: number
  account_id: string
  shop_name: string
  source: string
  region: string
  grain: string
  entity_type: string
  entity_id: string
  sku: string
  product_name: string
  impressions: number | null
  clicks: number | null
  add_to_cart: number | null
  orders: number | null
  impressions_rank: number | null
  clicks_rank: number | null
  add_to_cart_rank: number | null
  orders_rank: number | null
  focus_reasons: TrafficSkuFocusReason[]
}

export interface TrafficFilters {
  date_from: string
  date_to: string
  platform?: string[]
  platform_account_id?: number[]
  source?: string
  region?: string[]
}

export interface TrafficPeriodFallback {
  platform: string
  platform_account_id: number
  scope: 'current' | 'previous'
  requested_date_from: string
  requested_date_to: string
  actual_date_from: string
  actual_date_to: string
}

export interface TrafficRequestOptions {
  signal?: AbortSignal
}

function trafficQueryParams(params: object) {
  const searchParams = new URLSearchParams()
  Object.entries(params as Record<string, unknown>).forEach(([key, value]) => {
    if (value == null || value === '') return
    const values = Array.isArray(value) ? value : [value]
    values.forEach((item) => searchParams.append(key, String(item)))
  })
  return searchParams
}

export interface TrafficMetricResponse {
  items: TrafficMetricRow[]
  date_from: string
  date_to: string
  metric?: string
  sort_order?: TrafficSortOrder
  rank_scope?: 'global'
  fallback_periods?: TrafficPeriodFallback[]
}

export interface TrafficCategoryResponse {
  items: TrafficCategoryRow[]
  date_from: string
  date_to: string
  total_sku_count: number
  categorized_sku_count: number
  uncategorized_sku_count: number
  classification_rate: number
  fallback_periods?: TrafficPeriodFallback[]
}

export interface TrafficComparisonResponse {
  items: TrafficComparisonRow[]
  metric: string
  dimension: TrafficComparisonDimension
  sort_by: TrafficComparisonSort
  date_from: string
  date_to: string
  previous_date_from: string
  previous_date_to: string
  fallback_periods?: TrafficPeriodFallback[]
}

export interface TrafficSkuFocusResponse {
  items: TrafficSkuFocusRow[]
  top_n: number
  supported_metrics: TrafficMetricKey[]
  date_from: string
  date_to: string
  fallback_periods?: TrafficPeriodFallback[]
}

export interface TrafficCategorySkuComparisonFilters {
  date_from: string
  date_to: string
  metric: TrafficMetricKey
  sort_by?: TrafficComparisonSort
  keyword?: string
  change_direction?: TrafficComparisonChangeDirection
  limit?: number
  platform: string
  platform_account_id: number
  source: string
  grain: string
  region: string
  platform_category_id: string
}

export interface TrafficCategorySkuFocusFilters {
  date_from: string
  date_to: string
  top_n?: number
  keyword?: string
  platform: string
  platform_account_id: number
  source: string
  grain: string
  region: string
  platform_category_id: string
}

export function listTrafficAccounts(config?: { background?: boolean; silent?: boolean }) {
  return get<{ items: TrafficAccount[] }>('/api/v1/traffic-analytics/accounts', config)
}

export function fetchTrafficSummary(params: TrafficFilters, options?: TrafficRequestOptions) {
  return get<TrafficMetricResponse>('/api/v1/traffic-analytics/summary', {
    params: trafficQueryParams(params),
    background: true,
    signal: options?.signal
  })
}

export function fetchDailyNegativeReviews(params: TrafficFilters, options?: TrafficRequestOptions) {
  return get<TrafficMetricResponse>('/api/v1/traffic-analytics/negative-reviews', {
    params: trafficQueryParams(params),
    background: true,
    signal: options?.signal
  })
}

export function fetchTrafficCategories(params: TrafficFilters, options?: TrafficRequestOptions) {
  return get<TrafficCategoryResponse>('/api/v1/traffic-analytics/categories', {
    params: trafficQueryParams(params),
    background: true,
    signal: options?.signal
  })
}

export function fetchTrafficRankings(
  params: TrafficFilters & { metric: TrafficRankingMetricKey; sort_order?: TrafficSortOrder; limit?: number },
  options?: TrafficRequestOptions
) {
  return get<TrafficMetricResponse>('/api/v1/traffic-analytics/rankings', {
    params: trafficQueryParams(params),
    background: true,
    signal: options?.signal
  })
}

export function fetchTrafficComparison(
  params: TrafficFilters & {
    metric: TrafficMetricKey
    dimension?: TrafficComparisonDimension
    sort_by?: TrafficComparisonSort
    limit?: number
  },
  options?: TrafficRequestOptions
) {
  return get<TrafficComparisonResponse>('/api/v1/traffic-analytics/comparison', {
    params: trafficQueryParams(params),
    background: true,
    signal: options?.signal
  })
}

export function fetchTrafficCategorySkuComparison(
  params: TrafficCategorySkuComparisonFilters,
  options?: TrafficRequestOptions
) {
  return get<TrafficComparisonResponse>('/api/v1/traffic-analytics/comparison/category-skus', {
    params: trafficQueryParams(params),
    background: true,
    silent: true,
    signal: options?.signal
  })
}

export function fetchTrafficCategorySkuFocusAnalysis(
  params: TrafficCategorySkuFocusFilters,
  options?: TrafficRequestOptions
) {
  return get<TrafficSkuFocusResponse>('/api/v1/traffic-analytics/analysis/category-skus', {
    params: trafficQueryParams(params),
    background: true,
    silent: true,
    signal: options?.signal
  })
}

export function syncTrafficAnalytics(payload: {
  platform_account_ids?: number[]
  date_from: string
  date_to: string
}) {
  return post<{ items: TrafficSyncRun[] }>('/api/v1/traffic-analytics/sync', payload)
}

File: caifuclaw_business_app/app/api/routes/traffic_analytics.py
# Company: 深圳智柠网络科技有限公司
# Author: mohsen liang

from collections.abc import Callable
from datetime import date
from typing import Any

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from ...database import get_db
from ...models import LocalUser
from ...traffic_analytics import (
    TrafficSyncRequest,
    create_traffic_sync_runs,
    list_traffic_accounts,
    query_categories,
    query_category_sku_comparison,
    query_category_sku_focus_analysis,
    query_comparison,
    query_negative_reviews_daily,
    query_rankings,
    query_summary,
    run_traffic_sync_runs,
    validate_period,
)


def create_traffic_analytics_router(
    *,
    current_user_dependency: Callable[..., Any],
) -> APIRouter:
    router = APIRouter(prefix="/api/v1/traffic-analytics", tags=["traffic-analytics"])

    @router.get("/accounts")
    def traffic_analytics_accounts(
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        return {"items": list_traffic_accounts(db)}

    @router.post("/sync")
    def sync_traffic_analytics(
        payload: TrafficSyncRequest,
        background_tasks: BackgroundTasks,
        user: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            runs, pending_ids = create_traffic_sync_runs(db, payload, user.username)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        if pending_ids:
            background_tasks.add_task(run_traffic_sync_runs, pending_ids)
        return {"items": runs}

    @router.get("/summary")
    def traffic_analytics_summary(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        platform: list[str] | None = Query(None),
        platform_account_id: list[int] | None = Query(None),
        source: str = Query(""),
        region: list[str] | None = Query(None),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return query_summary(
            db,
            start,
            end,
            platform=platform or [],
            platform_account_id=platform_account_id,
            source=source.strip().lower(),
            region=region or [],
        )

    @router.get("/negative-reviews")
    def traffic_analytics_negative_reviews(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        platform: list[str] | None = Query(None),
        platform_account_id: list[int] | None = Query(None),
        source: str = Query(""),
        region: list[str] | None = Query(None),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return query_negative_reviews_daily(
            db,
            start,
            end,
            platform=platform or [],
            platform_account_id=platform_account_id,
            source=source.strip().lower(),
            region=region or [],
        )

    @router.get("/rankings")
    def traffic_analytics_rankings(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        metric: str = Query("impressions"),
        sort_order: str = Query("desc"),
        limit: int = Query(20, ge=1, le=100),
        platform: list[str] | None = Query(None),
        platform_account_id: list[int] | None = Query(None),
        source: str = Query(""),
        region: list[str] | None = Query(None),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
            return query_rankings(
                db,
                start,
                end,
                metric=metric,
                sort_order=sort_order,
                limit=limit,
                platform=platform or [],
                platform_account_id=platform_account_id,
                source=source.strip().lower(),
                region=region or [],
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc

    @router.get("/categories")
    def traffic_analytics_categories(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        platform: list[str] | None = Query(None),
        platform_account_id: list[int] | None = Query(None),
        source: str = Query(""),
        region: list[str] | None = Query(None),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return query_categories(
            db,
            start,
            end,
            platform=platform or [],
            platform_account_id=platform_account_id,
            source=source.strip().lower(),
            region=region or [],
        )

    @router.get("/comparison")
    def traffic_analytics_comparison(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        metric: str = Query("clicks"),
        dimension: str = Query("sku"),
        sort_by: str = Query("delta_abs"),
        limit: int = Query(20, ge=1, le=100),
        platform: list[str] | None = Query(None),
        platform_account_id: list[int] | None = Query(None),
        source: str = Query(""),
        region: list[str] | None = Query(None),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
            return query_comparison(
                db,
                start,
                end,
                metric=metric,
                dimension=dimension,
                sort_by=sort_by,
                limit=limit,
                platform=platform or [],
                platform_account_id=platform_account_id,
                source=source.strip().lower(),
                region=region or [],
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc

    @router.get("/comparison/category-skus")
    def traffic_analytics_category_sku_comparison(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        metric: str = Query("clicks"),
        sort_by: str = Query("delta_abs"),
        limit: int = Query(20, ge=1, le=100),
        platform: str = Query(""),
        platform_account_id: int = Query(0),
        source: str = Query(""),
        grain: str = Query(""),
        region: str = Query(""),
        platform_category_id: str = Query(""),
        keyword: str = Query(""),
        change_direction: str = Query("all"),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
            return query_category_sku_comparison(
                db,
                start,
                end,
                metric=metric,
                sort_by=sort_by,
                limit=limit,
                platform=platform,
                platform_account_id=platform_account_id,
                source=source,
                grain=grain,
                region=region,
                platform_category_id=platform_category_id,
                keyword=keyword,
                change_direction=change_direction,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc

    @router.get("/analysis/category-skus")
    def traffic_analytics_category_sku_focus_analysis(
        date_from: date | None = Query(None),
        date_to: date | None = Query(None),
        top_n: int = Query(20, ge=1, le=100),
        platform: str = Query(""),
        platform_account_id: int = Query(0),
        source: str = Query(""),
        grain: str = Query(""),
        region: str = Query(""),
        platform_category_id: str = Query(""),
        keyword: str = Query(""),
        _: LocalUser = Depends(current_user_dependency),
        db: Session = Depends(get_db),
    ) -> dict:
        try:
            start, end = validate_period(date_from, date_to)
            return query_category_sku_focus_analysis(
                db,
                start,
                end,
                top_n=top_n,
                platform=platform,
                platform_account_id=platform_account_id,
                source=source,
                grain=grain,
                region=region,
                platform_category_id=platform_category_id,
                keyword=keyword,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc

    return router
